From 9a9761d324843d50e3e9c2b84928ea1628fc7be6 Mon Sep 17 00:00:00 2001
From: Lin Cheng <lin_cheng@dell.com>
Date: Mon, 4 Sep 2023 10:51:56 +0000
Subject: [PATCH] src: nodejs root store add partial chain flag

---
 src/crypto/crypto_context.cc | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/src/crypto/crypto_context.cc b/src/crypto/crypto_context.cc
index 34ff222dbd..1e1d1d36aa 100644
--- a/src/crypto/crypto_context.cc
+++ b/src/crypto/crypto_context.cc
@@ -729,6 +729,9 @@ void SecureContext::SetCACert(const BIOPointer& bio) {
     CHECK_EQ(1, X509_STORE_add_cert(cert_store, x509.get()));
     CHECK_EQ(1, SSL_CTX_add_client_CA(ctx_.get(), x509.get()));
   }
+  if (cert_store) {
+    X509_STORE_set_flags(cert_store, X509_V_FLAG_PARTIAL_CHAIN);
+  }
 }
 
 void SecureContext::AddCACert(const FunctionCallbackInfo<Value>& args) {
@@ -786,6 +789,7 @@ void SecureContext::SetRootCerts() {
 
   // Increment reference count so global store is not deleted along with CTX.
   X509_STORE_up_ref(store);
+  X509_STORE_set_flags(store, X509_V_FLAG_PARTIAL_CHAIN);
   SSL_CTX_set_cert_store(ctx_.get(), store);
 }
 
-- 
2.25.1

